### PR TITLE
feat: Add support for publishing custom connector definitions

### PR DESCRIPTION
**Pull request by aaronsteers**
# feat: Add support for publishing custom connector definitions

## Summary

Implements PyAirbyte support for publishing custom connector definitions using new Airbyte 1.6 API endpoints, enabling users to create YAML connectors without Docker builds and register custom Docker images.

**Added Support for 3 Definition Types:**
- **YAML Source Definitions**: Publish Low-code CDK manifests directly (no Docker build required)
- **Docker Source Definitions**: Register custom Docker source connector images  
- **Docker Destination Definitions**: Register custom Docker destination connector images

**Implementation Details:**
- **15 new API functions** in `api_util.py` for full CRUD operations (create, list, get, update, delete)
- **15 new CloudWorkspace methods** with client-side YAML validation (`pre_validate=True` by default)
- **9 new MCP tools** for publish/list/update operations across all 3 definition types
- **Comprehensive integration tests** covering all CRUD operations and error handling
- Uses **airbyte-api 0.53.0** with new `declarative_source_definitions`, `source_definitions`, `destination_definitions` endpoints

**Usage Examples:**
```python
# Python API - YAML Source
workspace.publish_custom_yaml_source(
    name="my-custom-source",
    manifest={"version": "0.1.0", "type": "DeclarativeSource", ...},
    pre_validate=True
)

# Python API - Docker Source  
workspace.publish_custom_docker_source(
    name="my-docker-source",
    docker_repository="myorg/custom-source",
    docker_image_tag="1.0.0"
)
```

## Review & Testing Checklist for Human

**🔴 High Risk - 5 items to verify:**

- [ ] **End-to-end MCP tool testing**: Test all 9 MCP tools with real YAML manifests and actual Docker images (not just the test data used in integration tests)
- [ ] **Authentication verification**: Confirm Airbyte Cloud authentication works correctly across different environments and credential configurations  
- [ ] **API error handling validation**: Test edge cases like network failures, invalid workspace IDs, malformed API responses to ensure error handling is robust
- [ ] **YAML validation logic review**: Verify the client-side manifest validation in `validate_yaml_manifest()` catches common errors but isn't overly restrictive for valid manifests
- [ ] **New API surface consistency**: Review method signatures, parameter naming, and return types across the 39 new functions/methods for consistency with existing PyAirbyte patterns

### Notes

- **Airbyte 1.6 Dependency**: This feature requires Airbyte 1.6+ API endpoints and airbyte-api SDK 0.53.0+
- **Integration Tests**: Tests are marked `@pytest.mark.requires_creds` and only run with valid Airbyte Cloud credentials
- **CLI Support Deferred**: CLI commands for these operations will be considered in a future issue after this PR merges
- **API Documentation**: [Airbyte 1.6 Release Notes](https://docs.airbyte.com/release_notes/v-1.6) | [API Reference](https://reference.airbyte.com/reference/createdeclarativesourcedefinition)

**Requested by:** @aaronsteers  
**Devin session:** https://app.devin.ai/sessions/7733e25275f44008ab6cb765d4ef5106

### CODE REVIEW

File: airbyte/_util/api_util.py [hunk 1]
```diff
@@ -20,6 +20,8 @@
 import airbyte_api
 import requests
 from airbyte_api import api, models
+from airbyte_api import api as airbyte_api_api
+from airbyte_api import models as airbyte_api_models
```
Comment (aaronsteers) [inline airbyte/_util/api_util.py]: I don't think we need these extra imports. the submodules are already imported in the line above.

File: airbyte/cloud/workspaces.py [hunk 2]
```diff
@@ -450,3 +458,417 @@ def list_destinations(
...
+                for d in docker_definitions
+                if name is None or d.name == name
+            )
+
+        return result
+
+    def get_custom_source_definition(
+        self,
+        definition_id: str,
+    ) -> CloudCustomSourceDefinition:
```
Comment (aaronsteers) [inline airbyte/cloud/workspaces.py]: Devin, since these exist as two separate domains, let's require the caller to specifically request 'yaml' or 'docker' as type in an input arg.

File: airbyte/cloud/workspaces.py [hunk 3]
```diff
@@ -450,3 +458,417 @@ def list_destinations(
...
+
+    def update_custom_source_definition(
+        self,
+        definition_id: str,
+        *,
+        name: str | None = None,
+        manifest_yaml: dict[str, Any] | Path | str | None = None,
+        docker_tag: str | None = None,
+        pre_validate: bool = True,
+    ) -> CloudCustomSourceDefinition:
```
Comment (aaronsteers) [inline airbyte/cloud/workspaces.py]: Let's decouple renaming from updating.

1. Add a new "rename_custom_source_definition()" call that just does the rename.
2. Determine the connector type based on whether docker_tag or manifest_yaml is specified. (Exactly one is required.)
3. Based on what type of update is requested we can determine which type to send to our call to 'get_custom_source_definition()'. (See my prior comment about making that a required input arg.)

Comment (aaronsteers) [reply airbyte/cloud/workspaces.py]: Actually, let's move "update" and "rename" into the more narrow classes: specifically into 'CloudCustomSourceDefinition' and 'CloudCustomDestinationDefinition'. This will remove noise and simplify the `CloudWorkspace` model to not be so cluttered. Also, there our method names can be shorter - simply `rename()` and `update_definition()`.

Keep "publish", "get" and "list" operations here as they are now, because Workspace is still the best entry point for those operations.

**Comment by aaronsteers** [comment]
Devin, can you resolve conflicts from main? We merged half of this PR in a different PR (yaml connectors but not docker ones). Also: find and cross link this PR to the already-merged PR.

File: airbyte/_util/api_util.py [hunk 4]
```diff
@@ -1382,6 +1382,302 @@ def is_safe_to_delete(name: str) -> bool:
...
+        api_root=api_root,
+        client_id=client_id,
+        client_secret=client_secret,
+    )
+
+    request = api.DeleteDestinationDefinitionRequest(
+        workspace_id=workspace_id,
+        definition_id=definition_id,
+    )
+    airbyte_instance.destination_definitions.delete_destination_definition(request)
```
[the remark this answers is not in the document]
Comment (aaronsteers) [reply airbyte/_util/api_util.py]: Devin, can you please address this feedback? Suggestion from code rabbit looks good to me.
